Apply indent rule while typing.

rebornix opened the pull request:
Apply indent rule when users type. For instance, since not all languages uses brackets as block, some languages are using `decreaseIndentRule` to adjust indentation.

Take Ruby as example

![ruby-indentation](https://cloud.githubusercontent.com/assets/876920/20914835/fa4d0b96-bb36-11e6-9997-65292ef52774.gif)

rebornix commented:
@alexandrudima made updates to this PR.

Previously we calculate the indentation in two steps
* After getting the enter action, make updates to `enterAction.appendText` and `indentation` in `LanguageConfigurationRegistryImpl.getEnterActionAtPosition`.
* Calculate indentation again based on enterAction type in `TypeOperations._enter` or `TypeOperations._goodIndentForLine`.

This is a little bit confusing and easy to make mistake as both `appendText` and `indentation` can contain indentation information (`\t`), besides we are updating them in two places.

In current change, we only calculate `enterAction` in `LanguageConfigurationRegistryImpl`  and do all indentation calculation in `_enter` or `_goodIndentForLine`.

alexandrudima commented on src/vs/editor/common/modes/supports/onEnter.ts:
```diff
@@ -78,20 +78,32 @@ export class OnEnterSupport {
+					if (enterAction === null) {
+						enterAction = { indentAction: IndentAction.None, outdentCurrentLine: true };
+					} else {
+						enterAction.outdentCurrentLine = true;
```
Given this line modifies enterAction, please remove the instances OnEnterSupport._INDENT, OnEnterSupport._OUTDENT and create new object instances everywhere, otherwise the second call to this function might contain a stale outdentCurrentLine set to true.

alexandrudima commented on src/vs/editor/common/controller/cursorTypeOperations.ts:
```diff
@@ -194,40 +219,53 @@ export class TypeOperations {
-		let indentation = r.indentation;
+		// adjust indentation of current line.
+		let beforeText = '';
+		if (enterAction.outdentCurrentLine) {
```
When running scripts\test.bat --coverage this block is not covered, can you please write some tests that cover this new functionality?

alexandrudima commented:
@rebornix I have notived the following regression while testing this branch:

* indent with spaces @ 4
* press Enter after a closing bracket in js
* two issues: indentation is removed from the line holding the closing bracket and it is substituted with tabs (this is a hint that normalizeIndentation is not called somewhere?)

![regression1](https://cloud.githubusercontent.com/assets/5047891/21497225/3d2f16a6-cc2c-11e6-991e-d82bad3c6714.gif)

rebornix commented:
@alexandrudima addressed your comments above and added test cases for newly added code. About the regression you mentioned, there are two issues behind it:

1. After we adjust the indentation of current line, the new indentation is not normalized. It's fixed in latest commit now.
2. Since we are interpreting `decreaseIndentPattern` in a more reasonable way, the indentation rules used in TypeScript extension are out-of-date/wrong now. They followed the old semantics of `decreaseIndentPattern` so it can't work well with latest code. However, as the indentation rules used in TypeScript are just about brackets, we can remove them and let it fall back to our logic of brackets handling. Our brackets handling can make sure the indentation is correct. @dbaeumer @mjbvz you may want to take a look at this change and let me if you have concern about it.

dbaeumer commented:
Very old code :-) If it still works now I have no objection.

alexandrudima commented:
@rebornix In this case, we need to introduce a new concept, we cannot expect all extensions to update their `decreaseIndentPattern` to work around this semantic breaking change. Moreover, I do not agree that the `decreaseIndentPattern` contributed by TypeScript is wrong in any way: 

https://manual.macromates.com/en/appendix gives exactly brackets as an example
![image](https://cloud.githubusercontent.com/assets/5047891/21545291/222fa866-cddf-11e6-87ce-fb3354b2ae1b.png)

The root problem here is: if `decreaseIndentPattern` matches a line, it means that line needs to be indented at one indentation level "smaller" than the previous line. There is no code in the PR to check that, it currently decides to decrease the indentation level of a matching line, even if the indentation level of a matching line is already correct.

Let's discuss this today.

rebornix commented on src/vs/editor/test/common/controller/cursor.test.ts:
```diff
@@ -2146,64 +2156,6 @@ suite('Editor Controller - Regression tests', () => {
 		});
 	});
 
-	test('issue Microsoft/monaco-editor#108 part 1/2: Auto indentation on Enter with selection is half broken', () => {
```
Move it to indentation related test suite and uses indent rules instead of mocking `enterAction`.

alexandrudima commented:
@rebornix 

Some more cases that have issues:

### (High) pressing enter inserts tabs in this case (JSconfigured with spaces@4)
```js
function t() {

            if (a) {
                function bar() {
                    console.log('hello world');
                    switch(a) {
                        case 3:
                            console.log('hi');
                            console.log('bye');
                            
                    }
                }
            }|

}
```
![mixed-whitespace](https://cloud.githubusercontent.com/assets/5047891/22027400/10c39ae8-dcd4-11e6-8eb0-f63a764bd9cb.gif)

### (High) Languages without indent patterns part 1 (PHP with spaces@4)
* not how php does not define any indent patterns. 
* somehow the new logic sneaks in even in cases when a language does not have indent patterns defined
* same issue exists in C++
```php
<?

    if (true) {
        echo $hi;
        echo $bye;
    }|

?>
```

![jumpy-php](https://cloud.githubusercontent.com/assets/5047891/22027680/e1ed06a4-dcd4-11e6-9abf-bac65bb02c10.gif)

### (High) Languages without indent patterns part 2 (C++ with spaces@4)
```cpp
    if (5)
        return 5;
    |
```
actual:
![cpp-actual](https://cloud.githubusercontent.com/assets/5047891/22027920/b20c2838-dcd5-11e6-8cf4-17be625e93f0.gif)

expected:
![cpp-expected](https://cloud.githubusercontent.com/assets/5047891/22027927/b750245c-dcd5-11e6-828f-c5ff30f000cd.gif)

rebornix commented:
@alexandrudima I've fixed the four issues you mentioned. The reason for that is I'm handling the indentation of current line too eagerly. I tried to make the indentation of current line correct if possible. So original mechanism is likely

* if there is no indentation rule or there is no matched indentation rule, the expected indentation for current line inherits from previous lines.
* If beforeEnterText matches `decreaseIndentPattern`, decrease from the expected indentation rule.

That's why you can see we are always aligning current line with above lines. But right now I took one step back, only when current line's indentation is the same as the expected indentation and current line matches `decreaseIndentPattern`, we adjust current line's indentation. Otherwise we just use current line's indentation as baseline because people might change the indentation of current line by purpose (or maybe Code does intentionally, by `ElectricChar` logic).

I can see there are still some corner cases that this engine doesn't work properly, especially when there is a `decreaseIndentPattern` out there.

rebornix commented:
I've tried my best but the code is no longer rebase-able somehow (git just refuses to rebase) so I have to merge with master :(

alexandrudima commented:
Good job! :100:

alexandrudima commented:
@rebornix Remember to create a test plan item for this.
